gossipsub v1.1: introduce message signing policy (#294)
libp2p/specs: pubsub/gossipsub/gossipsub-v1.1.md

}
```

### Signature policy

The usage of the `signature`, `key`, `from`, and `seqno` fields in `Message` is now configurable.
These fields may negatively affect privacy in content-addressed messaging,

> and may need to be strictly enforced in author-addressed messaging.

Reviewer: I think the content-addressed and author-addressed word choice might be a bit confusing. How about "anonymous" and "origin-stamped"?

In gossipsub v1.0, a "lax" signing policy is effective: verify signatures, and if not, only when present.
In gossipsub v1.1, these fields are strictly present and verified, or completely omitted altogether.
An implementation may choose to support the legacy v1.0 "lax" signing policy,

>  along with an explicit message authoring option. 

Reviewer: "origin-stamping" option.

Reviewer: Move this paragraph below the specification of both options. Otherwise we're speaking about v1.0, then v1.1, then v1.0 support in v1.1, back to v1.1.

Author: Sounds good, can you include this in the changes you wanted to make? I'm including this line mostly for awareness of backwards compatibility options, but it does not need to be part of the spec otherwise.

Gossipsub v1.1 has two policies to choose from:
- `StrictSign`:
  - Produces the `signature`, `key` (`from` may be enough), `from` and `seqno` fields.
  - Enforce the fields to be present, reject otherwise.
  - Verify the signature, reject otherwise.
  - Propagate the fields if valid.
- `StrictNoSign`:
  - Produces messages without the `signature`, `key`, `from` and `seqno` fields.

>     The corresponding protobuf key-value pairs are absent from the marshalled message, not just empty.

Reviewer: Separate these in two sub-bullets. "On the message producer side:" / "On the message receiver side:"

}
```

> ### Signature Policy

Reviewer 2: I am not sure that this should be in the gossipsub spec -- it is a generic pubsub policy that applies to all routing algorithms.

Reviewer: Agreed, will move it.

Author: Was looking for a place to version it, but since it's not breaking anything necessarily, and applies to pubsub, moving it to the pubsub doc sounds good.

> The usage of the `signature`, `key`, `from`, and `seqno` fields in `Message` is now configurable.

Reviewer: ```suggestion
The usage of the `signature`, `key`, `from`, and `seqno` fields in `Message` is now configurable per topic, in the manners specified in this section.
```

> Initially this could be configured globally, however, configuration on a per-topic basis will facilitate mixed protocols better.

Reviewer 3: ```suggestion
Initially this could be configured globally, however, because the policies are mutually incompatible, configuration on a per-topic basis will facilitate mixed protocols better.
```

> Initially this could be configured globally, however, because the policies are mutually incompatible, configuration on a per-topic basis will facilitate mixed protocols better.

Reviewer: ```suggestion
> [[ Implementation note ]]: At the time of writing this section, go-libp2p-pubsub (reference implementation of this spec) allows for configuring the signature policy at a global pubsub instance level. This needs to be pushed down to topic-level configuration. Other implementations are encouraged to support topic-level configuration, as the spec mandates.
```

In the default origin-stamped messaging, the fields need to be strictly enforced:
the `seqno` and `from` fields form the `message_id`, and should be verified to avoid `message_id` collisions.

In content-stamped messaging, the fields may negatively affect privacy:

> revealing the relationship between `data` and `from`/`seqno`.

Reviewer: I'm not sure what the goal of this is. This might be better moved to after "Signature Policy Options", and put under a "Usage indications/notes" section.

I would reword to something like this:

* Origin-stamped messaging: choose `StrictSign`, and use in combination with the default `message_id_fn`.
* Content-addressed messaging: enable `StrictNoSign`, with a custom hash-based `message_id_fn`. Any signatures would now be part of the payload, and would need to be validated through a custom message validator.

Author: The idea was to give some context/intuition why the fields are handled like they are. It could be worded directly as `message id approach -> signing policy` directions, like you suggest. Either option works for us.

#### Signature Policy Options

In gossipsub v1.1, these fields are strictly present and verified, or completely omitted altogether:
- `StrictSign`:
  - On the producing side:

>     - Build messages with the `signature`, `key` (`from` may be enough), `from` and `seqno` fields.

Reviewer: ```suggestion
    - Build messages with the `signature`, `key` (`from` may be enough for certain inlineable public key types), `from` and `seqno` fields.
```

  - On the consuming side:
    - Enforce the fields to be present, reject otherwise.
    - Propagate only if the fields are valid and signature can be verified, reject otherwise.
- `StrictNoSign`:
  - On the producing side:
    - Build messages without the `signature`, `key`, `from` and `seqno` fields.
    - The corresponding protobuf key-value pairs are absent from the marshalled message, not just empty.
  - On the consuming side:
    - Enforce the fields to be absent, reject otherwise.
    - Propagate only if the fields are absent, reject otherwise.

>   - A `message_id` function will not be able to use the above fields, and may instead rely on the `data` field.

Reviewer: ```suggestion
  - A `message_id` function will not be able to use the above fields, and should instead rely on the `data` field. A commonplace strategy is to calculate a hash.
```

> In gossipsub v1.0, a legacy "lax" signing policy could be configured, to not verify signatures except when present:

Reviewer: ```suggestion
In gossipsub v1.0, a legacy "lax" signing policy could be configured, to only verify signatures when present. For security reasons, this is strategy is discarded in subsequent versions, but MAY still be supported for backwards-compatibility. If so, its use should be discouraged through prominent deprecation warnings. These strategies will be entirely dropped in the future.
```

> - `LaxSign`: *Defined for completeness, insecure*. Also known as authoring but not verifying.

Reviewer: ```suggestion
- `LaxSign`: *this was never an original gossipsub 1.0 option, but it's defined here for completeness, and considered insecure*. Always sign, and verify incoming signatures, and but accept unsigned messages.
```

  - On the producing side:
    - Build messages with the `signature`, `key` (`from` may be enough), `from` and `seqno` fields.
  - On the consuming side:
    - `signature` may be absent, and not verified.
    - Verify `signature`, iff the `signature` is present, then reject if `signature` is invalid.

> - `LaxNoSign`: *Previous default for no-verification*

Reviewer: ```suggestion
- `LaxNoSign`: *Previous default for no-verification*. Do not sign nor origin-stamp, but verify incoming signatures, and accept unsigned messages.
```
